cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -766,6 +806,7 @@ class CastKernel : public UnaryKernel {
   bool is_zero_copy_;
   bool can_pre_allocate_values_;
   std::shared_ptr<DataType> out_type_;
+  std::vector<std::unique_ptr<UnaryKernel>> child_cast_kernels;
```

Comment:
I'm sort of thinking of having a `ListCastKernel : public UnaryKernel` implementation. Would you mind if I took a pass at working on this?

Reply:
@wesm IMO, we don't have to create the special case like `ListCastKernel`. I think simply checking child `Field` is sufficient.

Reply:
You can take a look at what I just did -- for the moment casting sliced `ListArray` raises an exception, we'll need to do some more work to make that work

Reply:
I think we can generalize to support struct casts and have instead a `NestedCastKernel` when the time comes. I think we should defer that a bit until we've dealt with the sliced list case. The problem with the current `CastKernel` is that it's tied up in the business of memory allocation for primitive arrays. We'll keep needing to do some refactoring around this as we implement more kernels which are capable of writing into pre-allocated memory